### PR TITLE
Initialization rework

### PR DESCRIPTION
**Pull request by marktwtn**
refactor: Unify local and remote implementation
feat: Use other hardwares if remote interface fails to initialize
docs: Change the description of dcurl APIs

### CODE REVIEW

File: src/dcurl.c [hunk 1]
```diff
@@ -33,9 +33,19 @@
 
 /* check whether dcurl is initialized */
 static bool isInitialized = false;
+static uint8_t initStatus = 0;
 static uv_sem_t notify;
 
+/* for checking whether the corresponding implementation is initialized */
+#define C_BIT 1
```
Comment (jserv) [inline src/dcurl.c]: Can these lines be rewritten with `enum` and shifts?

File: src/dcurl.c [hunk 2]
```diff
@@ -33,9 +33,19 @@
 
 /* check whether dcurl is initialized */
 static bool isInitialized = false;
+static uint8_t initStatus = 0;
```
Comment (jserv) [inline src/dcurl.c]: Ditto. Use `enum` for `initStatus`.

File: src/dcurl.c [hunk 3]
```diff
@@ -60,26 +70,44 @@ static uv_sem_t notify_remote;
...
 bool dcurl_init()
 {
-    bool ret = true;
+    bool ret = false;
 
 #if defined(ENABLE_AVX)
-    ret &= registerImplContext(&PoWAVX_Context);
+    if (registerImplContext(&PoWAVX_Context)) {
+        initStatus = initStatus | AVX_BIT;
+        ret |= true;
```
Comment (jserv) [inline src/dcurl.c:81]: Valid value `1` specified in  `enum` would be much better for understanding purpose.

File: src/dcurl.c (hunk 2 above)
Comment (marktwtn) [reply src/dcurl.c]: We might have an enum like
```
enum Status {
    C_BIT = 1,
    SSE_BIT = 1 << 1,
    AVX_BIT = 1 << 2,
    GPU_BIT = 1 << 3,
    ...
};
```
and allow the `enum` value to be assigned to the `uint8_t initStatus`?

Comment (jserv) [reply src/dcurl.c]: You can also use the statements like `1 U << 1`.

File: src/dcurl.c [hunk 4]
```diff
@@ -60,26 +70,44 @@ static uv_sem_t notify_remote;
...
+    if (registerImplContext(&PoWFPGAAccel_Context)) {
+        initStatus = initStatus | FPGA_BIT;
+        ret |= true;
+    }
 #endif
 
 #if defined(ENABLE_REMOTE)
-    ret &= initializeRemoteContext(&Remote_Context);
+    if (registerRemoteContext(&Remote_Context)) {
+        initStatus = initStatus | REMOTE_BIT;
```
Comment (jserv) [inline src/dcurl.c]: I would prefer the statement:
```
initStatus |= REMOTE_BIT;
```

It would be shorter and meaningful.

File: src/dcurl.c (hunk 1 above)
Comment (jserv) [reply src/dcurl.c]: How about changing the name scheme from `*_BIT` to `CAP_*`? Thus, we can typedef `capability_t` from these values.

File: src/dcurl.c (hunk 4 above)
Comment (marktwtn) [reply src/dcurl.c]: Yes, I've noticed that.
I'll revise it in the next forced push.

File: src/dcurl.c (hunk 1 above)
Comment (marktwtn) [reply src/dcurl.c]: Done.
I named the `enum` as
```
enum Capability {
    CAP_C = 1U,
    CAP_SSE = 1U << 1,
    ...
}
```

I'm ok with the name scheme `CAP_*`.

File: src/dcurl.c (hunk 2 above)
Comment (marktwtn) [reply src/dcurl.c]: Done.

File: src/dcurl.c (hunk 4 above)
Comment (marktwtn) [reply src/dcurl.c]: Done.

File: src/dcurl.c (hunk 3 above)
Comment (marktwtn) [reply src/dcurl.c:81]: Got it.

File: src/dcurl.c [hunk 5]
```diff
@@ -33,9 +33,21 @@
 
 /* check whether dcurl is initialized */
 static bool isInitialized = false;
+static uint8_t initStatus = 0;
 static uv_sem_t notify;
 
+/* for checking whether the corresponding implementation is initialized */
+enum Capability {
+    CAP_C = 1U,
```
Comment (jserv) [inline src/dcurl.c:37]: If new entry `CAP_NONE = 0` is specified, `initStatus` can be set to `CAP_NONE`, that would be more consistent.

File: src/dcurl.c [hunk 6]
```diff
@@ -33,9 +33,21 @@
 
 /* check whether dcurl is initialized */
 static bool isInitialized = false;
+static uint8_t initStatus = 0;
```
Comment (jserv) [inline src/dcurl.c]: How about renaming from `initStatus` to `runtime_caps` or `caps`?

File: src/dcurl.c [hunk 7]
```diff
@@ -93,14 +123,23 @@ void dcurl_destroy()
...
+    }
 #endif
 
     list_for_each (p, &IMPL_LIST) {
         impl = list_entry(p, ImplContext, list);
         destroyImplContext(impl);
         list_del(p);
     }
+
+    initStatus = 0;
```
Comment (jserv) [inline src/dcurl.c]: Ditto. Avoid using raw `0` for reseting.

File: src/remote_interface.h [hunk 8]
```diff
@@ -55,8 +56,12 @@ struct _remote_impl_context {
     int8_t *(*getPoWResult)(void *pow_ctx);
     PoW_Info (*getPoWInfo)(void *pow_ctx);
     bool (*freePoWContext)(RemoteImplContext *remote_ctx, void *pow_ctx);
+
+    /* Linked list */
+    struct list_head list;
```
Comment (jserv) [inline src/remote_interface.h]: `list` might not be an effective name.

File: src/dcurl.c (hunk 5 above)
Comment (marktwtn) [reply src/dcurl.c:37]: Agree. It would be more readable.

File: src/dcurl.c (hunk 6 above)
Comment (marktwtn) [reply src/dcurl.c]: I prefer `runtimeCaps`.

It reminds me to unify the variable naming with **Camel-Case** or others.

File: src/remote_interface.h (hunk 8 above)
Comment (marktwtn) [reply src/remote_interface.h]: I just copied it from the file `src/implcontext.h` for unification.

**Comment by jserv** [comment]
Add implementation-specific information in file `docs/remote-interface.md`.

File: src/remote_interface.h (hunk 8 above)
Comment (marktwtn) [reply src/remote_interface.h]: I would like to change it to
```
/* Node in linked list */
struct list_head node;
```

File: src/dcurl.c (hunk 5 above)
Comment (marktwtn) [reply src/dcurl.c:37]: Done.

File: src/dcurl.c (hunk 7 above)
Comment (marktwtn) [reply src/dcurl.c]: Done.

File: src/remote_interface.h (hunk 8 above)
Comment (marktwtn) [reply src/remote_interface.h]: Done.

File: src/dcurl.c (hunk 6 above)
Comment (marktwtn) [reply src/dcurl.c]: Done.

File: docs/remote-interface.md [hunk 9]
```diff
@@ -77,3 +77,17 @@ $ make BUILD_REMOTE=1 BUILD_DEBUG=1 check
 
 ## Requirements
 Remote interface requires RabbitMQ broker
+
+## Fallback mechanism
+If the remote interface is not working for some reason, the PoW calculation will be transfer to the local hardwares.\
```
Comment (jserv) [inline docs/remote-interface.md]: Fix the grammar. "be transfer" -> "be transferred"

File: docs/remote-interface.md [hunk 10]
```diff
@@ -77,3 +77,17 @@ $ make BUILD_REMOTE=1 BUILD_DEBUG=1 check
...
+The possible situations are:
+
+* Initialization failure\
```
Comment (jserv) [inline docs/remote-interface.md]: Why trailing `\` character?

Comment (marktwtn) [reply docs/remote-interface.md]: The `\` character is for line break.

File: docs/remote-interface.md (hunk 9 above)
Comment (marktwtn) [reply docs/remote-interface.md]: Fixed.

File: docs/remote-interface.md [hunk 11]
```diff
@@ -77,3 +77,17 @@ $ make BUILD_REMOTE=1 BUILD_DEBUG=1 check
 
 ## Requirements
 Remote interface requires RabbitMQ broker
+
+## Fallback mechanism
+If the remote interface is not working for some reason, the PoW calculation will be transferred to the local hardwares.\
```
Comment (jserv) [inline docs/remote-interface.md]: I have no clue why you would have to add `\` manually? Markdown viewer shall be always responsible for appropriate layout.

Comment (marktwtn) [reply docs/remote-interface.md]: Sometimes I do not want the rendered line to be too long.
Hence I break the line manually for a better visualization in my point of view.

Shall I remove them all?
I added it in the other documents, too.

Comment (marktwtn) [reply docs/remote-interface.md]: If so, I would like to remove them in the new pull request, which is for removing `\` in the documents and revise the source code to fit the `clang format`.

Comment (jserv) [reply docs/remote-interface.md]: Contents are more important than formatting. Go ahead.

Comment (marktwtn) [reply docs/remote-interface.md]: I'll remove the `\` symbol added in this pull request.
The rest of them in other documents will be removed in the next new pull request.

Comment (marktwtn) [reply docs/remote-interface.md]: Done.

**Pull request merged**